compiler/pragmas.nim

```diff
@@ -1308,7 +1308,8 @@ proc singlePragma(c: PContext, sym: PSym, n: PNode, i: var int,
         noVal(c, it)
         if sym == nil: invalidPragma(c, it)
         else: sym.flags.incl sfUsed
-      of wLiftLocals: discard
+      of wLiftLocals:
+        sym.flags.incl(sfForceLift)
```

Comment:
Why is this required?

Reply:
The yasync library injects some variables into the closureiter body and expects they will be available in its env. After this pr this is no longer guaranteed. Hence I've added an escape hatch in form of `{.liftLocals.}` pragma for yasync to take advantage of it. I hope to find a better way in the future, but that likely will happen after state-split algorithm changes. As a side note you might be better off by reviewing just the commit "[liftLocals pragma for closureiter locals to be forcefully lifted" to better see the affected parts.

Reply:
Well it's not documented and if your library needs this annotation it implies bad things:

1. This is not 100% backwards compatible. (However, the change is beneficial enough to warrant it.)
2. This is not a safe transformation.

What does yasync do that requires the new annotation?

Reply:
is this new pragma documented?

Reply:
It has never been documented in the first place. `liftLocals` was invented to facilitate some experiments in the library/async/cps space, which are now abandoned and it is pretty much unused by anyone, correct me if I'm wrong. So I just gave this pragma another breath to be useful in yasync, I really don't expect anyone else to use it, just like it has been this way up to this point :)

> This is not 100% backwards compatible

Why not? It is backwards-compatible because previously it was allowed only for functions (and the old behavior remains). Now it is also allowed for variables with slightly different-but-close-enough meaning. If you're concerned that it will be removed or changed in the future, I don't mind, we'll figure it out when time comes.

> This is not a safe transformation.

Why not? You can place `{.liftLocals.}` willy-nilly on any variable and worst harm it will do is lift said local to the env, just like every other local was lifted until this pr.

> What does yasync do that requires the new annotation?

I don't know how detailed answer you're expecting, but yasync implements async and Futures by collapsing all the futures and closureiter environments in the callstack into one big environment which allows running async tasks with a single allocation per whole task, which can also happen on the stack in certain scenarios. For that to happen yasync injects the Continuation header, result storage and arguments into the closureiter code as variables, and later expects to find them lifted in the corresponding environment.